npm/pkg/dataplane/dpshim/dpshim.go

```diff
@@ -1,77 +1,590 @@
+		}
+
+		set := dp.setCache[prefixedSetName]
+		if set.IPSetMetadata.GetSetKind() != ipsets.HashSet {
+			return npmerrors.Errorf(npmerrors.AppendIPSet, false, fmt.Sprintf("RemoveFromSets, ipset %s is not a hash set", prefixedSetName))
+		}
+
+		// in case the IP belongs to a new Pod, then ignore this Delete call as this might be stale
+		cachedPod, exists := set.IPPodMetadata[podMetadata.PodIP]
+		if !exists {
+			continue
+		}
+		if cachedPod.PodKey != podMetadata.PodKey {
+			klog.Infof("DeleteFromSet: PodOwner has changed for Ip: %s, setName:%s, Old podKey: %s, new podKey: %s. Ignore the delete as this is stale update",
+				cachedPod.PodIP, prefixedSetName, cachedPod.PodKey, podMetadata.PodKey)
+			continue
+		}
+
+		// update the IP ownership with podkey
+		delete(set.IPPodMetadata, podMetadata.PodIP)
+		dp.dirtyCache.modifyAddorUpdateSets(prefixedSetName)
+	}
 	return nil
 }
 
-func (dp *DPShim) RemoveFromList(listName *ipsets.IPSetMetadata, setNames []*ipsets.IPSetMetadata) error {
+func (dp *DPShim) AddToLists(listMetadatas, setMetadatas []*ipsets.IPSetMetadata) error {
+	dp.Lock()
+	defer dp.Unlock()
+
+	if err := dp.checkForListMemberUpdateErrors(listMetadatas, setMetadatas, npmerrors.AppendIPSet); err != nil {
+		return err
+	}
+
+	for _, listMetadata := range listMetadatas {
+		listName := listMetadata.GetPrefixName()
+		list := dp.setCache[listName]
+		for _, setMetadata := range setMetadatas {
+			setName := setMetadata.GetPrefixName()
+			set := dp.setCache[setName]
+
+			if _, ok := list.MemberIPSets[setName]; ok {
+				continue
+			}
+			list.MemberIPSets[setName] = set.IPSetMetadata
+			set.AddReference(listName, controlplane.ListReference)
+			dp.dirtyCache.modifyAddorUpdateSets(setName)
+		}
+		dp.dirtyCache.modifyAddorUpdateSets(listName)
+	}
+
 	return nil
 }
 
-func (dp *DPShim) ApplyDataPlane() error {
+func (dp *DPShim) RemoveFromList(listMetadata *ipsets.IPSetMetadata, setMetadatas []*ipsets.IPSetMetadata) error {
+	dp.Lock()
+	defer dp.Unlock()
+
+	if err := dp.checkForListMemberUpdateErrors([]*ipsets.IPSetMetadata{listMetadata}, setMetadatas, npmerrors.DeleteIPSet); err != nil {
+		return err
+	}
+
+	listName := listMetadata.GetPrefixName()
+	list := dp.setCache[listName]
+	for _, setMetadata := range setMetadatas {
+		setName := setMetadata.GetPrefixName()
+		set := dp.setCache[setName]
+
+		if _, ok := list.MemberIPSets[setName]; !ok {
+			continue
+		}
+
+		delete(list.MemberIPSets, setName)
+		set.DeleteReference(listName, controlplane.ListReference)
+	}
+	dp.dirtyCache.modifyAddorUpdateSets(listName)
+
 	return nil
 }
 
 func (dp *DPShim) AddPolicy(networkpolicies *policies.NPMNetworkPolicy) error {
-	return nil
+	var err error
+	// apply dataplane after syncing
+	defer func() {
+		dperr := dp.ApplyDataPlane()
+		if dperr != nil {
+			err = fmt.Errorf("failed with error %w, apply failed with %v", err, dperr)
+		}
+	}()
+	dp.Lock()
+	defer dp.Unlock()
+
+	if dp.policyExists(networkpolicies.PolicyKey) {
```

Comment:
For another PR: what do we think of normalizing/validating policies before they're sent to dp? See these functions: https://github.com/Azure/azure-container-networking/blob/35dcb32b18dff9683e79007729713892acb21781/npm/pkg/dataplane/policies/policymanager.go#L146